### ISSUE TITLE
French language does not recognize plural strings

### ISSUE THREAD

**Issue by Kingdutch** [WeblateOrg/weblate#3119]
**Describe the bug**
For some reason Weblate thinks that plural strings are singular for the French language.

![afbeelding](https://user-images.githubusercontent.com/327697/67013767-ef0f7e80-f0f3-11e9-9772-c627d6bc1b25.png)

Even though it properly recognizes that the string is plural in other languages.

![afbeelding](https://user-images.githubusercontent.com/327697/67013871-1ebe8680-f0f4-11e9-9cd1-a97920f0312b.png)

The language file in git contains the proper source strings and translation.

[drupal-core-fr.txt](https://github.com/WeblateOrg/weblate/files/3739576/drupal-core-fr.txt) (extension changed from .po to .txt so it could be uploaded).

The downloaded file from Weblate also only contains the singular string. The database confirms the behaviour:

![afbeelding](https://user-images.githubusercontent.com/327697/67014038-6f35e400-f0f4-11e9-8114-3ea3a171c28a.png)

The language is also configured correctly.

![afbeelding](https://user-images.githubusercontent.com/327697/67014130-98567480-f0f4-11e9-944b-e1ba0725b797.png)

**To Reproduce**

I'm honestly not sure. Ideally I'd just like to wipe the French language in Weblate and force it to re-import everything from the *-fr.po files in git.

**Expected behavior**
I expected French to display the proper translations as if the strings were plural.

**Server configuration and status**
```
$ weblate list_versions
 * Weblate 3.5.1
 * Python 3.6.9
 * Django 2.1.7
 * Celery 4.2.2
 * celery-batches 0.2
 * six 1.12.0
 * social-auth-core 3.1.0
 * social-auth-app-django 3.1.0
 * django-appconf 1.0.3
 * translate-toolkit 2.3.1
 * translation-finder 1.1
 * Whoosh 2.7.4
 * defusedxml 0.5.0
 * Git 2.11.0
 * Pillow 5.4.1
 * python-dateutil 2.8.0
 * lxml 4.3.2
 * django-crispy-forms 1.7.2
 * django_compressor 2.2
 * djangorestframework 3.9.2
 * user-agents 1.1.0
 * jellyfish 0.7.1
 * diff-match-patch 20121119
 * pytz 2018.9
 * PyYAML 5.1
 * Mercurial 3.1.2
 * Database backends: django.db.backends.postgresql
 * Cache backends: avatar:FileBasedCache, default:RedisCache
 * Celery: redis://redis.internal:6379/1, redis://redis.internal:6379/1, regular
 * Platform: Linux 4.19.60-pl1 (x86_64)
```

```
$ weblate check --deploy

WARNINGS:
?: (security.W004) You have not set a value for the SECURE_HSTS_SECONDS setting. If your entire site is served only over SSL, you may want to consider setting a value and enabling HTTP Strict Transport Security. Be sure to read the documentation first; enabling HSTS carelessly can cause serious, irreversible problems.

INFOS:
?: (weblate.I021) Error collection is not configured, it is highly recommended for production use
	HINT: https://docs.weblate.org/en/weblate-3.5.1/admin/install.html#collecting-errors
```

**Comment by nijel** [comment]
Does the French PO file define the string as plural (does it have `msgid_plural`)?

**Comment by nijel** [comment]
Can you please attach the complete original PO file here? Or send to michal@cihar.com in case it's not something you want to make publicly available. 

You might also want look into Weblate repository history when it was changed.

**Comment by Kingdutch** [comment]
The complete PO file as it's in the git repository for the component was already attached to the original story (here: https://github.com/WeblateOrg/weblate/files/3739576/drupal-core-fr.txt )

Here is the file downloaded directly from Weblate: [open-social-drupal-core-fr.txt](https://github.com/WeblateOrg/weblate/files/3740049/open-social-drupal-core-fr.txt)

(I've changed the extensions from .po to .txt so I can upload it to GitHub. The contents is otherwise unchanged).

The history was cut-off from the original screenshot. Here is one that includes the history.

![afbeelding](https://user-images.githubusercontent.com/327697/67022928-4583b980-f102-11e9-82b0-1c7f3d30534b.png)

This is the git history for the git .po file:
```
a029d851 (Alexander Varwijk         2019-06-26 17:22:46 +0200  1185) msgid "1 year"
1b669dd6 (Jenkins Automation Server 2019-07-31 11:41:46 +0000  1186) msgid_plural "@count years"
1b669dd6 (Jenkins Automation Server 2019-07-31 11:41:46 +0000  1187) msgstr[0] "@count année"
1b669dd6 (Jenkins Automation Server 2019-07-31 11:41:46 +0000  1188) msgstr[1] "@count année"
```

So I do see it was changed on 2019-07-31 🤔 

The relevant diff of that commit is this:
```
#: includes/common.inc:0 lib/Drupal/Core/Datetime/DateFormatter.php:240
+#, fuzzy
 msgid "1 year"
-msgstr "@count année"
+msgid_plural "@count years"
+msgstr[0] "@count année"
+msgstr[1] "@count année"
```

The commit for `1 year` is the following:
```
git log -n1 a029d851
commit a029d85151da8d379d6f1ff0718d75222a92f827
Author: Alexander Varwijk <alexander@goalgorilla.com>
Date:   Wed Jun 26 17:22:46 2019 +0200

    Add translation files for 'fr' langcode.
```

Which has the following diff
```
+#: includes/common.inc:0 lib/Drupal/Core/Datetime/DateFormatter.php:240
+msgid "1 year"
+msgstr "@count année"
+
```

So it looks like it went wrong for France there for some reason (it's been fine for languages that were added using the msginit and msgmerge tool outside of Weblate both before and after France was added).

I guess at this point the question is, why is Weblate not throwing away the singular `1 year` string which is no longer in the .po file and not in the .pot file and pick up the plural version of it.
I've already tried running `weblate loadpo --all --force --lang fr` but this has not resolved the issue.

**Comment by nijel** [comment]
Maybe just the Weblate repository was not yet updated to see the latest version? Or there were some errors during the update so that it was not properly processed (check celery logs for errors)?

**Comment by Kingdutch** [comment]
I don't see any errors in the logs. Unfortunately I don't have logs from 3 months ago so I can't check whether it had an error at the time of the change.

We run both `updategit` and `commitgit` nightly so I would expect the Weblate repository to be up to date.

Checking the `data/vcs/open-social/drupal-core/` folder and looking in the `drupal-core-fr.po` file also shows that the file is up to date. Can I check somewhere the last time Weblate processed this file in the database?

**Comment by nijel** [comment]
The `revision` field on `trans_translation` should match hash returned by `git ls-tree HEAD drupal-core-fr.po`. In case it doesn't, the current file content hasn't been parsed.

PS: Looking at version you are using, it's possible that some errors are silently dropped in Celery, this has been fixed in 3.8 (see 417091902309b28e77e27ce2e4e845b596efa34a, #2813 and #2850).

**Comment by Kingdutch** [comment]
Thanks for the info! Unfortunately I'm not at the office right now but I'll look into this on Monday.

Unfortunately we're blocked from updating at the moment due to a missing server dependency that the hosting provider can not yet resolve 😞

**Comment by Kingdutch** [comment]
I just checked. The last revision for `core/drupal-core-fr.po` (full path in the repo to the file) is `b42dc6def74badd3f7afe7d6b6e14cd07f2be880`. I can see that this commit is not part of the history of the file in the git repository. Upon further investigation it looks like this commit doesn't exist in the git repository at all for any file.

**Comment by nijel** [comment]
It's not a commit, it's an object hash. You can get it by `git ls-tree HEAD core/drupal-core-fr.po` in the repo.

**Comment by Kingdutch** [comment]
Ah, learned something new :) 

Running the command that you provided gives the following output.

```
100644 blob b42dc6def74badd3f7afe7d6b6e14cd07f2be880	core/drupal-core-fr.po
```

**Comment by nijel** [comment]
Hmm, I was hoping it's not up to date, what would explain the behavior you see.

Reading again the previous comments, few things are a bit unclear to me:

* Is the `core/drupal-core-fr.po` file in the Weblate working repository correct?
* The above PO file downloaded from Weblate was downloaded using download original or through converted download?

**Comment by Kingdutch** [comment]
- Yes the copy of the file in the Weblate repository contains only 1 occurence of the string `1 year` which is the following.
```
#: includes/common.inc:0 lib/Drupal/Core/Datetime/DateFormatter.php:240
#, fuzzy
msgid "1 year"
msgid_plural "@count years"
msgstr[0] "@count année"
msgstr[1] "@count année"
```

- The above file was downloaded using the `Download translation file as gettext PO`

![afbeelding](https://user-images.githubusercontent.com/327697/67364802-8c453980-f570-11e9-9cd5-d6dbd5776596.png)

Here is the file downloaded using the `Download original translation file (Gettext PO)`: [open-social-drupal-core-fr-original.txt](https://github.com/WeblateOrg/weblate/files/3760585/open-social-drupal-core-fr-original.txt) (.po changed to .txt so it could be uploaded here)

**Comment by nijel** [comment]
Okay, that file is generated based on what Weblate has in the database, so it matches the screenshots in https://github.com/WeblateOrg/weblate/issues/3119#issue-508485563. I guess the original translation file looks correct.

So in the end it's problem that the updated file was not properly parsed for whatever reason. Can you try running `./manage.py loadpo --foreground --force --lang fr open-social/drupal-core` (I'm just guessing the URL slugs for the project/component, so please adjust accordingly).

**Comment by Kingdutch** [comment]
You guessed the component names correctly :) 

We don't have a `manage.py` file on our platform. Commands are run using the `weblate` command. However, this gives me the following:

```
$ weblate loadpo --foreground --force --lang fr open-social/drupal-core
usage: weblate loadpo [-h] [--all] [--lang LANG] [--force] [--version]
                      [-v {0,1,2,3}] [--settings SETTINGS]
                      [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                      [component [component ...]]
weblate loadpo: error: unrecognized arguments: --foreground
```

```
$ which weblate
/app/.global/bin/weblate
```

```
$ ls -al /app/.global/bin/weblate
-rwxr-xr-x 1 web web 229 Sep  3 11:06 /app/.global/bin/weblate
```

Searching for `manage.py` provides the following (hiding folders we're not allowed to access):
```
$ find / -name manage.py 2>/dev/null
/app/.global/lib/python3.6/site-packages/weblate/trans/views/manage.py
```

However, I don't think that's the CLI tool you want me to use :)

**Comment by nijel** [comment]
Ah, it's not yet present in your version, it was added later. I think patch adding it should apply cleanly on older version as well: 3c0ca80a07dbadb4826445fc9fdd28f95ba2704e

**Comment by Kingdutch** [comment]
Hi Stalebot ;-), sorry for that, did not yet have time to patch yet to test this.

**Comment by Kingdutch** [comment]
Yesterday we were able to make some infrastructure changes that make it easier for us to deploy changes to Weblate (unfortunately still missing the library that's preventing us from upgrading). This means that I was able to deploy the patch for the `--foreground` flag.

However, I see on production now that this issue has resolved itself. The revision for the language file is updated: `64f097345c3fcd46392b75423b39bb93788113ae`. On a sync that I did before deploying these changes to a development branch I can still see the old revision `b42dc6def74badd3f7afe7d6b6e14cd07f2be880` and there the issue persists.

It could be that the .po file was never properly updated on production because celery workers may have gotten killed before getting to them. On the production environment the redis queue currently contains 57382 items after receiving the latest string updates last night (on the develop branch this is 0 but it does not receive the changes).

The main point of the infrastructure changes was to run the workers in separate containers. This should ensure that they don't starve one another of resources, so I hope the queue goes down over the coming days. If that was indeed the cause then this should resolve the issue.

I've run the `weblate loadpo --foreground --force --lang fr open-social/drupal-core` on the develop environment to see if we can still figure out if this was indeed in our set-up or something else that the changes have now masked on deployment.

I can happily report that the command ran without any output (indicating success(?)) and that looking in the database afterwards I can see that the string is now properly sourced (and translated) as plural! 

That means this issue is solved :) Thanks for all the help in debugging this!
